File: elixir/test/symphony_elixir/knowledge_base_test.exs
defmodule SymphonyElixir.KnowledgeBaseTest do
  use ExUnit.Case, async: false

  alias SymphonyElixir.KnowledgeBase
  alias SymphonyElixir.LocalTracker.Context
  alias SymphonyElixir.Repo

  setup do
    migrate_repo()
    clean_repo()
    root = configure_isolated_workspace_root()

    {:ok, _project} =
      Context.create_workspace_project(%{
        "name" => "Acme",
        "slug" => "acme",
        "tracker" => %{"kind" => "local"},
        "repositories" => [
          %{"github_full_name" => "acme/web", "workspace_path" => "web", "role" => "frontend"}
        ],
        "setup" => %{}
      })

    checkout = Path.join([root, "acme", "web"])
    File.mkdir_p!(Path.join(checkout, "docs/architecture"))
    File.write!(Path.join(checkout, "docs/index.md"), "---\ntitle: Home\n---\n# Home\n")

    File.write!(
      Path.join(checkout, "docs/architecture/backend.md"),
      "---\ntitle: Backend\n---\n# B\n\nbody\n"
    )

    File.write!(Path.join(checkout, "docs/broken.md"), "---\n- not\n- a map\n---\nx")
    git(checkout, ["init", "-q", "-b", "main"])
    git(checkout, ["add", "-A"])
    git(checkout, ["-c", "user.email=t@t", "-c", "user.name=t", "commit", "-q", "-m", "seed docs"])
    {:ok, root: root, checkout: checkout}
  end

  defp git(dir, args), do: {_o, 0} = System.cmd("git", args, cd: dir, stderr_to_stdout: true)

  test "project_overview returns repositories with docs status" do
    assert {:ok, overview} = KnowledgeBase.project_overview("acme")
    assert overview.project.slug == "acme"
    assert [%{repo_slug: "web", docs_present?: true}] = overview.repositories
  end

  test "project_overview returns error for unknown project" do
    assert KnowledgeBase.project_overview("nope") == {:error, :project_not_found}
  end

  test "repo_tree returns the repository summary and tree" do
    assert {:ok, result} = KnowledgeBase.repo_tree("acme", "web")
    assert result.repository.repo_slug == "web"
    assert result.docs_present == true
    assert Enum.any?(result.tree, &(&1.name == "index.md"))
  end

  test "repo_tree errors for unknown repo and project" do
    assert KnowledgeBase.repo_tree("acme", "missing") == {:error, :repo_not_found}
    assert KnowledgeBase.repo_tree("nope", "web") == {:error, :project_not_found}
  end

  test "read_page returns frontmatter, title, body, and content" do
    assert {:ok, page} = KnowledgeBase.read_page("acme", "web", ["architecture", "backend.md"])
    assert page.title == "Backend"
    assert page.path == "architecture/backend.md"
    assert page.frontmatter["title"] == "Backend"
    assert page.body =~ "body"
    assert page.content =~ "# B"
  end

  test "read_page validates path and missing files" do
    assert KnowledgeBase.read_page("acme", "web", ["..", "x.md"]) == {:error, :kb_invalid_path}
    assert KnowledgeBase.read_page("acme", "web", ["nope.md"]) == {:error, :kb_page_not_found}
    assert KnowledgeBase.read_page("acme", "web", ["broken.md"]) == {:error, :kb_frontmatter_invalid}
  end

  test "a saved page is immediately findable via project search" do
    {:ok, _} =
      KnowledgeBase.write_page("acme", "web", "search-me.md", %{
        frontmatter: %{"title" => "Find Me"},
        body: "a unique zebra phrase"
      })

    assert {:ok, results} = KnowledgeBase.search_project("acme", "zebra", [])
    assert Enum.any?(results, &(&1.path == "search-me.md"))
    assert Enum.all?(results, &(&1.repo_slug == "web"))
  end

  test "docs committed on disk become searchable after the repo tree is loaded" do
    # The seed commits docs to disk without ever calling write_page, so the
    # search index starts empty for this repo (the bug: only UI-authored pages
    # were indexed, leaving Git-committed docs unsearchable).
    assert {:ok, []} = KnowledgeBase.search_project("acme", "Backend", [])

    # Loading the repo tree is what the KB UI does on open; it must reindex the
    # repository's docs from disk so every committed page becomes findable.
    assert {:ok, _tree} = KnowledgeBase.repo_tree("acme", "web")

    assert {:ok, results} = KnowledgeBase.search_project("acme", "Backend", [])
    assert Enum.any?(results, &(&1.path == "architecture/backend.md"))
    assert Enum.all?(results, &(&1.repo_slug == "web"))
  end

  test "deleting a saved page removes it from search results" do
    {:ok, _} =
      KnowledgeBase.write_page("acme", "web", "temp.md", %{
        frontmatter: %{"title" => "Temp"},
        body: "ephemeral wombat note"
      })

    assert {:ok, [_ | _]} = KnowledgeBase.search_project("acme", "wombat", [])
    {:ok, _} = KnowledgeBase.delete_page("acme", "web", "temp.md")
    assert {:ok, []} = KnowledgeBase.search_project("acme", "wombat", [])
  end

  test "sync_status returns idle for a never-synced repo" do
    assert {:ok, %{status: "idle"}} = KnowledgeBase.sync_status("acme", "web")
  end

  test "sync helpers are inert no-ops for project KB base checkouts" do
    assert :ok = KnowledgeBase.request_sync("acme", "web")

    assert {:ok, %{status: "idle", last_error: nil, pr_number: nil, pr_url: nil}} =
             KnowledgeBase.sync_status("acme", "web")
  end

  describe "personal KB (@user scope)" do
    @general_repo_slug "@user~symphony-kb"

    setup %{root: root} do
      origin = Path.join(root, "kb-origin")
      File.mkdir_p!(Path.join(origin, "docs"))
      File.write!(Path.join(origin, "docs/keep.md"), "---\ntitle: Keep\n---\n# Keep\n")
      git(origin, ["init", "-q", "-b", "main"])
      git(origin, ["add", "-A"])
      git(origin, ["-c", "user.email=t@t", "-c", "user.name=t", "commit", "-q", "-m", "seed"])

      deps = [
        ensure_repo: fn ->
          {:ok,
           %{
             full_name: "octocat/symphony-kb",
             clone_url: origin,
             default_branch: "main",
             created: false
           }}
        end,
        clone: fn _clone_url, dest ->
          {_o, 0} = System.cmd("git", ["clone", "-q", origin, dest], stderr_to_stdout: true)
          {:ok, dest}
        end
      ]

      Application.put_env(:symphony_elixir, :kb_general_deps, deps)
      on_exit(fn -> Application.delete_env(:symphony_elixir, :kb_general_deps) end)
      :ok
    end

    test "project_overview reports the synthetic Personal repo, disconnected before connect" do
      assert {:ok, overview} = KnowledgeBase.project_overview("@user")
      assert overview.project.slug == "@user"
      assert [%{repo_slug: @general_repo_slug, docs_present?: false}] = overview.repositories

      assert {:ok, _} = KnowledgeBase.general_connect()

      assert {:ok, connected} = KnowledgeBase.project_overview("@user")
      assert [%{repo_slug: @general_repo_slug, docs_present?: true}] = connected.repositories
    end

    test "repo_tree and read_page resolve against the personal checkout" do
      assert {:ok, _} = KnowledgeBase.general_connect()

      assert {:ok, tree} = KnowledgeBase.repo_tree("@user", @general_repo_slug)
      assert tree.repository.repo_slug == @general_repo_slug
      assert tree.docs_present == true
      assert Enum.any?(tree.tree, &(&1.name == "keep.md"))

      assert {:ok, page} = KnowledgeBase.read_page("@user", @general_repo_slug, ["keep.md"])
      assert page.title == "Keep"
      assert page.repo_slug == @general_repo_slug
    end

    test "a saved personal page is findable via @user search" do
      assert {:ok, _} = KnowledgeBase.general_connect()

      {:ok, _} =
        KnowledgeBase.general_write_page("notes/zebra.md", %{
          frontmatter: %{"title" => "Z"},
          body: "a unique zebra phrase"
        })

      assert {:ok, results} = KnowledgeBase.search_project("@user", "zebra", [])
      assert Enum.any?(results, &(&1.path == "notes/zebra.md"))
    end

    test "sync helpers are inert no-ops for the personal KB" do
      assert :ok = KnowledgeBase.request_sync("@user", @general_repo_slug)
      assert {:ok, %{status: "idle"}} = KnowledgeBase.sync_status("@user", @general_repo_slug)
    end
  end

  defp migrate_repo do
    {:ok, _repo, _apps} =
      Ecto.Migrator.with_repo(Repo, fn repo -> Ecto.Migrator.run(repo, :up, all: true) end)
  end

  defp clean_repo, do: SymphonyElixir.TestSupport.truncate_tracker!(Repo)

  defp configure_isolated_workspace_root do
    root = Path.join(System.tmp_dir!(), "kb-context-#{System.unique_integer([:positive])}")
    File.mkdir_p!(root)
    workflow = Path.join(root, "WORKFLOW.md")
    SymphonyElixir.TestSupport.write_workflow_file!(workflow, workspace_root: root)
    SymphonyElixir.Workflow.set_workflow_file_path(workflow)

    on_exit(fn ->
      File.rm_rf(root)
      Application.delete_env(:symphony_elixir, :workflow_file_path)
    end)

    root
  end
end

File: tracker/src/services/__tests__/knowledgeBase.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import { http } from "@/services/http";
import {
  connectGeneral,
  getGeneralOverview,
  getPage,
  getProjectOverview,
  getRepoTree,
  getSyncStatus,
  regenerateGeneralHome,
  requestSync,
  savePage,
  searchProject,
} from "@/services/knowledgeBase";

vi.mock("@/services/http", async (orig) => {
  const actual = await orig<typeof import("@/services/http")>();
  return {
    ...actual,
    http: { get: vi.fn(), put: vi.fn(), post: vi.fn(), delete: vi.fn() },
  };
});

afterEach(() => vi.clearAllMocks());

describe("knowledgeBaseService", () => {
  it("getProjectOverview maps repositories (docs_present? key + workspace path)", async () => {
    (http.get as ReturnType<typeof vi.fn>).mockResolvedValue({
      data: {
        data: {
          project: { slug: "acme", name: "Acme" },
          repositories: [
            {
              repo_slug: "services~api",
              workspace_path: "services/api",
              github_full_name: "acme/api",
              role: "backend",
              "docs_present?": true,
            },
          ],
        },
      },
    });

    const overview = await getProjectOverview("acme");
    expect(http.get).toHaveBeenCalledWith("/api/tracker/v1/projects/acme/kb");
    expect(overview.project.name).toBe("Acme");
    expect(overview.repositories[0].repoSlug).toBe("services~api");
    expect(overview.repositories[0].workspacePath).toBe("services/api");
    expect(overview.repositories[0].docsPresent).toBe(true);
  });

  it("getRepoTree requests the repo path (no /tree suffix) and uses top-level docs_present", async () => {
    (http.get as ReturnType<typeof vi.fn>).mockResolvedValue({
      data: {
        data: {
          repository: {
            repo_slug: "web",
            workspace_path: "web",
            github_full_name: "acme/web",
            role: "frontend",
          },
          docs_present: true,
          tree: [],
        },
      },
    });

    const result = await getRepoTree("acme", "web");
    expect(http.get).toHaveBeenCalledWith("/api/tracker/v1/projects/acme/kb/repos/web");
    expect(result.docsPresent).toBe(true);
    expect(result.repository.docsPresent).toBe(true);
  });

  it("getPage maps backend `content` to `markdown`", async () => {
    (http.get as ReturnType<typeof vi.fn>).mockResolvedValue({
      data: {
        data: { repo_slug: "web", path: "a.md", title: "A", frontmatter: {}, body: "b", content: "# A\n\nb" },
      },
    });

    const page = await getPage("acme", "web", "a.md");
    expect(http.get).toHaveBeenCalledWith("/api/tracker/v1/projects/acme/kb/repos/web/pages/a.md");
    expect(page.markdown).toBe("# A\n\nb");
    expect(page.title).toBe("A");
  });

  it("savePage PUTs frontmatter + body and maps the result", async () => {
    (http.put as ReturnType<typeof vi.fn>).mockResolvedValue({
      data: { data: { path: "a.md", commit: "abc", pushed: false } },
    });

    const result = await savePage("acme", "web", "a.md", { frontmatter: { title: "A" }, body: "# A" });
    expect(http.put).toHaveBeenCalledWith("/api/tracker/v1/projects/acme/kb/repos/web/pages/a.md", {
      frontmatter: { title: "A" },
      body: "# A",
    });
    expect(result.commit).toBe("abc");
    expect(result.pushed).toBe(false);
  });

  it("searchProject passes q and repo params", async () => {
    (http.get as ReturnType<typeof vi.fn>).mockResolvedValue({ data: { data: [] } });
    await searchProject("acme", "query", { repo: "web" });
    expect(http.get).toHaveBeenCalledWith("/api/tracker/v1/projects/acme/kb/search", {
      params: { q: "query", repo: "web" },
    });
  });

  it("getSyncStatus and requestSync hit the sync endpoints", async () => {
    (http.get as ReturnType<typeof vi.fn>).mockResolvedValue({
      data: { data: { status: "idle", pr_number: null, pr_url: null, last_error: null, last_synced_at: null } },
    });
    (http.post as ReturnType<typeof vi.fn>).mockResolvedValue({ data: { data: { accepted: true } } });

    const state = await getSyncStatus("acme", "web");
    await requestSync("acme", "web");
    expect(http.get).toHaveBeenCalledWith("/api/tracker/v1/projects/acme/kb/repos/web/sync");
    expect(http.post).toHaveBeenCalledWith("/api/tracker/v1/projects/acme/kb/repos/web/sync");
    expect(state.status).toBe("idle");
    expect(state.prNumber).toBeNull();
  });

  it("general KB endpoints map connection + tree", async () => {
    (http.get as ReturnType<typeof vi.fn>).mockResolvedValue({
      data: { data: { connected: true, tree: [{ type: "page", name: "index.md", path: "index.md", title: "Home", order: null, favorite: false, children: [] }] } },
    });
    (http.post as ReturnType<typeof vi.fn>).mockResolvedValue({ data: { data: { connected: true } } });

    const overview = await getGeneralOverview();
    await connectGeneral();
    expect(http.get).toHaveBeenCalledWith("/api/tracker/v1/kb");
    expect(http.post).toHaveBeenCalledWith("/api/tracker/v1/kb/connect");
    expect(overview.connected).toBe(true);
    expect(overview.tree[0].path).toBe("index.md");

    (http.post as ReturnType<typeof vi.fn>).mockResolvedValue({ data: { data: { path: "index.md", commit: "x", pushed: false } } });
    const result = await regenerateGeneralHome();
    expect(http.post).toHaveBeenCalledWith("/api/tracker/v1/kb/home");
    expect(result.path).toBe("index.md");
  });
});
